### ISSUE TITLE
Infer static member references

### ISSUE THREAD

**Issue by jez9999** [Microsoft/TypeScript#19450]
In the following code:

```
export default class SettingsParser {
	private static _settings: Settings;

	/** Parse commandline, check settings, and init logging to appropriate loglevel */
	public static parse(settings: Settings) {
		_settings = settings;
	}
}
```

The `_settings` reference is flagged as an error with the message "[ts] Cannot find name '_settings'. Did you mean the static member 'SettingsParser._settings'?"  Yes, I did.  C# just infers this.  Why can't TypeScript?  Having to write `SettingsParser.` each time is very verbose.

**Comment by andy-ms** [comment]
TypeScript just type-checks JavaScript code -- it doesn't alter it. Since your example won't work when run as JavaScript, TypeScript will issue an error.

**Comment by jez9999** [comment]
It alters it a bunch, like making `import` statements into `require` statements if necessary.  Why not add the class name when it is useful?  It doesn't work run as JS but that's not a design goal of TS - the design goal is "valid JavaScript is valid TypeScript", not the other way around.

**Comment by DanielRosenwasser** [comment]
See https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

**Comment by kitsonk** [comment]
> It alters it a bunch, like making `import` statements into `require` statements if necessary.

That is transforming one form of valid TypeScript to another form.  Transpiling.

**Comment by jez9999** [comment]
That's somewhat splitting hairs.  It's changing variable names.  Fundamentally it is changing the input code.  Why does it matter if it isn't valid JavaScript going in?

**Comment by jez9999** [comment]
@DanielRosenwasser So if it's a "bad idea" to emit this different code, what would one suggest as a good alternative?  Do we really have to write out the whole class name every time?

**Comment by andy-ms** [comment]
You could just use a `let` in the scope of the class.
```ts
let settings: Settings;
export default class SettingsParser {
	public static parse(settings: Settings) {
		_settings = settings;
	}
}
```

**Comment by jez9999** [comment]
@andy-ms You mean `let _settings: Settings;`?  Is that logically the same as a static member, though?

**Comment by andy-ms** [comment]
It's not the same as a static member since it's not a property on the class -- it can only be accessed in its scope. But since you wanted the property to be `private` anyway, that's probably what you want.

**Comment by DanielRosenwasser** [comment]
If you need that, there's

```
let _settings = {};
export default class SettingsParser {
    get settings() {
        return _settings;
    }
}
```

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
